File: code/chapter5/normal_mle.py
from scipy.optimize import minimize
from math import log

def log_lik(theta, x):
  mu = theta[0]
  sigma = theta[1]
  # return the negative log-likelihood for minimization
  return -(-len(x)*log(sigma) - 0.5/sigma**2*sum((e-mu)**2 for e in x))

def normal_mle(x):
  return minimize(fun=lambda e:log_lik(e,x),x0=[0,1],method='L-BFGS-B',bounds=((None,None),(1e-6,None))).x

File: code/chapter5/normal_mle.R
log_lik = function(theta, x){
  mu = theta[1]
  sigma = theta[2]
  # we return the negative log-likelihood since optim is for minimization
  -(-length(x)*log(sigma) - 0.5/sigma^2*sum((x-mu)^2))
}

normal_mle = function(x){
  # par=c(0,1) as the initial values
  # lower=c(-Inf, 1e-6) for the lower bounds
  optim(par=c(0, 1), fn=log_lik, x=x, lower=c(-Inf, 1e-6), method="L-BFGS-B")$par
}

